### DISCUSSION TITLE
Threading for WebSocket

### DISCUSSION THREAD

**Discussion by xarthurx** [ocornut/imgui#3919] [Old/Archived discussions]
I've read multiple issues relating to threading for non-OpenGL process but am still not very clear if `imgui` can do this or not.
So sorry for the spam here.

**Need**:
In my UI, I have two buttons: `startServer` and `stopServer`, which will start/stop a WebSocket Server.
The Server will receive data from time to time and send things back after processing the data.

The main UI can be blocked or not blocked after starting the Server, but I would like to have the functionality to press `stopServer`.
Meanwhile, after I press the `startServer`, the main UI is frozen after the WebSockets started.

From what I understand, I need to put it in another thread. 
But does `imgui` allows this? I'm not vary familiar with GUI and would like to know if the instant gui (`imgui`) can do this.

![image](https://user-images.githubusercontent.com/1921878/111179858-c02d5e80-85ac-11eb-8cc3-57f6084547a8.png)

**Comment by nicolasnoble** [comment]
Here, imgui wouldn't really have any sort of interaction with your threaded code. Your C/C++ code is responsible for being the threadsafe one.

Something like this:

```c++

if (ImGui::Button("Start Server")) {
    threadSafeEventStartServer();
}

ImGui::SameLine();

if (ImGui::Button("Stop Server")) {
    threadSafeEventStopServer();
}
```

You are responsible for writing `theradSafeEventStartServer()` and `threadSafeEventStopServer()` in a thread safe manner that is.

**Reply by xarthurx** [reply]
Thank you!

I'm just confused if ImGui can do this or not after reading some of the issues.
Since it can, I'll handle the rest part.
